server/chat-plugins/chatlog.ts

```diff
@@ -166,10 +165,69 @@ export const LogReader = new class {
 		const prevMonth = new Date(new Date(`${month}-15`).getTime() - 30 * DAY);
 		return prevMonth.toISOString().slice(0, 7);
 	}
-
 	today() {
 		return Chat.toTimestamp(new Date()).slice(0, 10);
 	}
+	compareDates(a: string, b: string) {
+		const aDate = new Date(a).getTime();
+		const bDate = new Date(b).getTime();
+		return bDate - aDate;
+	}
+	isDate(item: string) {
+		return !isNaN(new Date(item).getTime());
+	}
+	async getBattleLog(tier: ID, number: number) {
+		// binary search!
+		const months = (await FS('logs').readdir())
+			.filter(f => this.isDate(f) && FS(`logs/${f}/${tier}`).existsSync())
+			.sort(this.compareDates);
+		if (!months.length) return;
+
+		const getBattleNum = (battleName: string) => Number(battleName.split('-')[1].slice(0, -9));
+
+		const getMonthRange = (month: string) => {
+			const files = FS(`logs/${month}/${tier}/`).readdirSync();
+			const firstDay = files[0];
+			const firstPath = FS(`logs/${month}/${tier}/${firstDay}`);
+			const firstFiles = firstPath.readdirSync();
+			if (files.length === 1) {
+				const num = getBattleNum(firstFiles[0]);
+				return [num, num];
+			}
+			const lastDay = files[files.length - 1];
+
+			const secondPath = FS(`logs/${month}/${tier}/${lastDay}`);
+			const firstResult = getBattleNum(firstFiles[0]);
+			const lastFiles = secondPath.readdirSync();
+			const secondResult = getBattleNum(lastFiles[lastFiles.length - 1]);
+			return [firstResult, secondResult];
+		};
+
+		while (months.length) {
+			const mIndex = Math.round(months.length / 2);
+			let middle = months[mIndex];
+			if (!middle) {
+				// one month left
+				middle = months.pop()!;
+			}
+			const [lowest, highest] = getMonthRange(middle);
+
+			if (number < lowest) {
+				months.splice(0, mIndex);
+			} else if (number > highest) {
+				months.splice(mIndex);
+			} else { // in that month
+				const days = (await FS(`logs/${middle}/${tier}`).readdir()).filter(f => this.isDate(f));
+				for (const day of days) {
```

Comment:
Binary search inside the days too!

Reply:
You don't actually need to treat days and months separately, just grab the first day of the first month, the last day of the last month, and interpolate.

Reply:
As I said on discord, why do it inside the days? You have all the other info needed to find the file, why not just iterate and see if it's there?

Reply:
As I said on Discord, your method makes the code more complicated than it needs to be, in addition to also being slower. There's no reason for you to treat days and months separately.

Reply:
Touche. Speaking of, check discord please - sent some relevant stuff there.